crates/stackable-versioned-macros/src/codegen/module.rs

```diff
@@ -1,53 +1,114 @@
 
-pub(crate) struct ModuleInput {
-    pub(crate) vis: Visibility,
-    pub(crate) ident: Ident,
-}
-
 /// A versioned module.
 ///
 /// Versioned modules allow versioning multiple containers at once without introducing conflicting
 /// version module definitions.
-pub(crate) struct Module {
+pub struct Module {
     versions: Vec<VersionDefinition>,
+
+    // Recognized items of the module
     containers: Vec<Container>,
-    preserve_module: bool,
-    skip_from: bool,
+    submodules: Vec<Submodule>,
+
     ident: IdentString,
     vis: Visibility,
+
+    // Flags which influence generation
+    preserve_module: bool,
+    skip_from: bool,
 }
 
 impl Module {
     /// Creates a new versioned module containing versioned containers.
-    pub(crate) fn new(
-        ModuleInput { ident, vis, .. }: ModuleInput,
-        preserve_module: bool,
-        skip_from: bool,
-        versions: Vec<VersionDefinition>,
-        containers: Vec<Container>,
-    ) -> Self {
-        Self {
-            ident: ident.into(),
+    pub fn new(item_mod: ItemMod, module_attributes: ModuleAttributes) -> Result<Self> {
+        let Some((_, items)) = item_mod.content else {
+            return Err(
+                Error::custom("the macro can only be used on module blocks").with_span(&item_mod)
+            );
+        };
+
+        let versions: Vec<VersionDefinition> = (&module_attributes).into();
+
+        let preserve_module = module_attributes
+            .common
+            .options
+            .preserve_module
+            .is_present();
+
+        let skip_from = module_attributes
+            .common
+            .options
+            .skip
+            .as_ref()
+            .map_or(false, |opts| opts.from.is_present());
+
+        let mut errors = Error::accumulator();
+        let mut containers = Vec::new();
+        let mut submodules = Vec::new();
+
+        for item in items {
+            match item {
+                Item::Enum(item_enum) => {
+                    let container = Container::new_enum_nested(item_enum, &versions)?;
+                    containers.push(container);
+                }
+                Item::Struct(item_struct) => {
+                    let container = Container::new_struct_nested(item_struct, &versions)?;
+                    containers.push(container);
+                }
+                Item::Mod(item_module) => match item_module.content {
+                    Some((_, items)) => {
+                        let imports: Vec<ItemUse> = items
+                            .into_iter()
+                            // We are only interested in use statements. Everything else is ignored.
+                            // NOTE (@Techassi): We might want to error instead.
+                            .filter_map(|item| match item {
+                                Item::Use(item_use) => Some(item_use),
+                                _ => None,
```

Comment:
I think this should be an error, suggesting they do that outside.

Reply:
Done in 12b9883